### PR TITLE
5746 admin page design concept and implementation

### LINKED ISSUE #5746: React Admin High Level Architecture Document
## Background ##

The Django admin currently serves as the primary administrative interface for TDP. While functional, it is tightly coupled to the Django ORM, difficult to customize for our domain-specific workflows, and provides a disjointed user experience compared to the TDP React-based frontend. This issue tracks the creation of a high-level architecture document for replacing the Django admin with a React-based admin console.

The architecture document should evaluate and recommend an approach -- specifically whether to build the admin console as a **separate Next.js application** leveraging server-side rendering (SSR) and React Server Components (RSC) for handling large tabular datasets, complex filtering, and server-driven data workflows that are core to admin operations.

### Why It Matters ###

- The Django admin lacks the flexibility to support custom admin workflows (Reparsing is a good example)
- A React-based admin console would unify the tech stack, making it easier for the team to maintain and extend both user-facing and admin-facing surfaces
- Server-side rendering via Next.js would address performance concerns around rendering large datasets (e.g., parsed record tables, parser errors) that are impractical to handle entirely client-side
- Building in React allows us to leverage USWDS components, which are designed to meet Section 508 accessibility requirements out of the box. The Django admin has no built-in USWDS support, making 508 compliance costly to achieve and maintain.
- This aligns with the longer-term goal of decoupling the frontend presentation layer from Django

### What's Changing ###

This issue produces an **architecture document only** -- no UI design, wireframes, or visual mockups. The deliverable is a technical recommendation covering application structure, rendering strategy, form strategy, data access patterns, and deployment considerations.

### Acceptance Criteria ###

- [ ] Document evaluates building the admin console as a standalone Next.js app vs. integrating admin routes into the existing frontend
- [ ] Document addresses the use of React Server Components and/or SSR for rendering large tabular data (parsed records, error reports, submission logs)
- [ ] Document covers API integration strategy -- whether the Next.js app communicates with the existing Django REST API or requires a BFF (Backend for Frontend) layer
- [ ] Document covers authentication and authorization, including how admin sessions and permissions are managed (e.g., reusing the existing Django auth/session infrastructure vs. token-based auth). Keycloak can be very helpful here if the implementer splits the admin into a Next.js or other separate style of app.
- [ ] Document covers deployment topology -- how the Next.js app (if used) is built, deployed, and served alongside the existing frontend and backend in the current Cloud.gov infrastructure
- [ ] Document addresses data access patterns for key admin workflows: user management, data file submission review, parsed record inspection, and error report viewing
- [ ] Document identifies migration strategy -- how to incrementally deprecate Django admin views as React admin equivalents are delivered
- [ ] Document includes a rough technology stack recommendation (Next.js version, key libraries for tables/data grids, state management approach)

### Usability Testing Criteria ###

_N/A -- this is a technical architecture document, not a user-facing deliverable._

### Notes ###

- The existing frontend is a Create React App (CRA) project using React 18, Redux, and React Router 6. The architecture document should consider whether adopting Next.js for the admin console creates an opportunity or a burden relative to this existing stack.
- This document is **not** prescribing UI/UX design. Visual design, component libraries, page layouts, and interaction patterns are out of scope and will be addressed in separate design issues.

### Open Questions ###

- Should the Next.js admin app (if suggested) share any code (components, utilities, API clients) with the existing CRA frontend, or should it be fully independent?
- What is the preferred strategy for handling large datasets -- server-side pagination via API, cursor-based streaming via RSC, or a hybrid approach?

### PR DESCRIPTION
**Pull request by raftmsohani**
## Summary of Changes
_Provide a brief summary of changes_
Pull request closes #5746 _

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
1. Proceed with functional tests as described herein.
1. Test steps should be captured in the demo GIF(s) and/or screenshots below.
> *Demo GIF(s) and screenshots for testing procedure*

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Document evaluates building the admin console as a standalone Next.js app vs. integrating admin routes into the existing frontend
- [x] Document addresses the use of React Server Components and/or SSR for rendering large tabular data (parsed records, error reports, submission logs)
- [x] Document covers API integration strategy -- whether the Next.js app communicates with the existing Django REST API or requires a BFF (Backend for Frontend) layer
- [x] Document covers authentication and authorization, including how admin sessions and permissions are managed (e.g., reusing the existing Django auth/session infrastructure vs. token-based auth). Keycloak can be very helpful here if the implementer splits the admin into a Next.js or other separate style of app.
- [x] Document covers deployment topology -- how the Next.js app (if used) is built, deployed, and served alongside the existing frontend and backend in the current Cloud.gov infrastructure
- [x] Document addresses data access patterns for key admin workflows: user management, data file submission review, parsed record inspection, and error report viewing
- [x] Document identifies migration strategy -- how to incrementally deprecate Django admin views as React admin equivalents are delivered
- [x] Document includes a rough technology stack recommendation (Next.js version, key libraries for tables/data grids, state management approach)
+ [ ] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

**Review by elipe17 — CHANGES_REQUESTED** [review]
I took a first pass at this and had some big points to address. My overarching suggestion for the doc and some extra points to consider are below. 


The document is trying to serve three purposes at once, which is the main driver of it feeling scattered and non-cohesive. It currently blends three distinct concerns:

   - **Architecture Decision Record (ADR):** The CRA vs Next.js comparison, decision matrix, rationale, and "Final Recommendation Rationale" section. This answers *"what did we choose and why?"*
   - **Architecture Specification:** The component architecture, auth strategy, data access patterns, deployment topology, rendering strategy, and migration strategy. This answers *"how is the system structured?"*
   - **Implementation Guide:** The TypeScript code snippets, directory scaffolds, and `manifest.yml` examples. This answers *"how do we build it?"*

   These three concerns have different audiences, different lifespans, and different levels of detail. Mixing them too much causes the reader to constantly shift between evaluating a decision, understanding a design, and reading implementation code.Some crossover is inherently required for important integration points/references, but the level of detail here is excessive. 

   The primary purpose should be the Architecture Specification. E.g. describing the structure, boundaries, and key design decisions of the Next.js admin console. The other two concerns could be handled as follows:

   - **ADR content (CRA vs Next.js):** Condense into a short section near the top of this document (problem statement → options considered → decision → brief rationale). The current ~450-line CRA analysis, bundle size comparison, Redux complexity example, and performance timeline diagrams are ADR-level justifications that can be trimmed to a snippet in this document. If we want to preserve the full breakdown for posterity it can be moved to a separate ADR document (Alex might appreciate this).
   - **Implementation guide content (code snippets, manifest.yml):** Remove prescriptive implementation from this document. Small code snippets or small configuration snippets are fine when they are in reference to a key integration point. But this document doesn't need to prescribe the middleware implementation on the Next.js app. Simply describing the structure (not implementation) is sufficient.

   This would leave the architecture document focused on the technical design of key components: Next.js, SSR, Generic Form builders/components, authentication, TDP integration, risks/open questions, etc...

Some other small points I had:
- Add the current CRA frontend to the diagram in the High-Level Component Architecture diagram section
- Remove Tailwind CSS from the UI & Styling section. USWDS is the styling library
- The "Open Questions" section answers its own questions with implementation code.

**Comment by raftmsohani** [comment]
> I took a first pass at this and had some big points to address. My overarching suggestion for the doc and some extra points to consider are below.
> 
> The document is trying to serve three purposes at once, which is the main driver of it feeling scattered and non-cohesive. It currently blends three distinct concerns:
> 
> * **Architecture Decision Record (ADR):** The CRA vs Next.js comparison, decision matrix, rationale, and "Final Recommendation Rationale" section. This answers _"what did we choose and why?"_
> * **Architecture Specification:** The component architecture, auth strategy, data access patterns, deployment topology, rendering strategy, and migration strategy. This answers _"how is the system structured?"_
> * **Implementation Guide:** The TypeScript code snippets, directory scaffolds, and `manifest.yml` examples. This answers _"how do we build it?"_
> 
> These three concerns have different audiences, different lifespans, and different levels of detail. Mixing them too much causes the reader to constantly shift between evaluating a decision, understanding a design, and reading implementation code.Some crossover is inherently required for important integration points/references, but the level of detail here is excessive.
> 
> The primary purpose should be the Architecture Specification. E.g. describing the structure, boundaries, and key design decisions of the Next.js admin console. The other two concerns could be handled as follows:
> 
> * **ADR content (CRA vs Next.js):** Condense into a short section near the top of this document (problem statement → options considered → decision → brief rationale). The current ~450-line CRA analysis, bundle size comparison, Redux complexity example, and performance timeline diagrams are ADR-level justifications that can be trimmed to a snippet in this document. If we want to preserve the full breakdown for posterity it can be moved to a separate ADR document (Alex might appreciate this).
> * **Implementation guide content (code snippets, manifest.yml):** Remove prescriptive implementation from this document. Small code snippets or small configuration snippets are fine when they are in reference to a key integration point. But this document doesn't need to prescribe the middleware implementation on the Next.js app. Simply describing the structure (not implementation) is sufficient.
> 
> This would leave the architecture document focused on the technical design of key components: Next.js, SSR, Generic Form builders/components, authentication, TDP integration, risks/open questions, etc...
> 
> Some other small points I had:
> 
> * Add the current CRA frontend to the diagram in the High-Level Component Architecture diagram section
> * Remove Tailwind CSS from the UI & Styling section. USWDS is the styling library
> * The "Open Questions" section answers its own questions with implementation code.

@elipe17 does it make sense to add two separate docs for adding more detailed domain diagrams and then another one for implementation suggestion or we are going one highlevel design for now and later add more detail to it for other audience?

**Comment by elipe17** [comment]
> > I took a first pass at this and had some big points to address. My overarching suggestion for the doc and some extra points to consider are below.
> > The document is trying to serve three purposes at once, which is the main driver of it feeling scattered and non-cohesive. It currently blends three distinct concerns:
> > 
> > * **Architecture Decision Record (ADR):** The CRA vs Next.js comparison, decision matrix, rationale, and "Final Recommendation Rationale" section. This answers _"what did we choose and why?"_
> > * **Architecture Specification:** The component architecture, auth strategy, data access patterns, deployment topology, rendering strategy, and migration strategy. This answers _"how is the system structured?"_
> > * **Implementation Guide:** The TypeScript code snippets, directory scaffolds, and `manifest.yml` examples. This answers _"how do we build it?"_
> > 
> > These three concerns have different audiences, different lifespans, and different levels of detail. Mixing them too much causes the reader to constantly shift between evaluating a decision, understanding a design, and reading implementation code.Some crossover is inherently required for important integration points/references, but the level of detail here is excessive.
> > The primary purpose should be the Architecture Specification. E.g. describing the structure, boundaries, and key design decisions of the Next.js admin console. The other two concerns could be handled as follows:
> > 
> > * **ADR content (CRA vs Next.js):** Condense into a short section near the top of this document (problem statement → options considered → decision → brief rationale). The current ~450-line CRA analysis, bundle size comparison, Redux complexity example, and performance timeline diagrams are ADR-level justifications that can be trimmed to a snippet in this document. If we want to preserve the full breakdown for posterity it can be moved to a separate ADR document (Alex might appreciate this).
> > * **Implementation guide content (code snippets, manifest.yml):** Remove prescriptive implementation from this document. Small code snippets or small configuration snippets are fine when they are in reference to a key integration point. But this document doesn't need to prescribe the middleware implementation on the Next.js app. Simply describing the structure (not implementation) is sufficient.
> > 
> > This would leave the architecture document focused on the technical design of key components: Next.js, SSR, Generic Form builders/components, authentication, TDP integration, risks/open questions, etc...
> > Some other small points I had:
> > 
> > * Add the current CRA frontend to the diagram in the High-Level Component Architecture diagram section
> > * Remove Tailwind CSS from the UI & Styling section. USWDS is the styling library
> > * The "Open Questions" section answers its own questions with implementation code.
> 
> @elipe17 does it make sense to add two separate docs for adding more detailed domain diagrams and then another one for implementation suggestion or we are going one highlevel design for now and later add more detail to it for other audience?

@raftmsohani my thinking would be that more detailed implementation style guides are unnecessary and that the design and the decision on the design is more important. Implementation is really cheap now with LLMs. The design and reason why are far more important. I think it makes sense to move the CRA vs Next.js analysis into a ADR. The ADR can also call out why it is nice to move away from DAC etc.... Then the document you have now can really spell out design patterns, structure, etc..... and remain really focused on that. Like I said in my comment, some of the info bleeds between all three styles of documents. So some small code examples in the architecture doc to make a point or clearly define a critical integration is perfectly ok.

**Review by jtimpe — COMMENTED** [review]
Looks good to me in general. One thing I didn't see addressed, though, was that Django provides a lot of convenience features through the existing admin form. Django models already have field validators; if we proceed with the implementation as you've outlined, we would need to duplicate these validators in order to have frontend form validation (pre-submission, think field types, ranges, etc). Have you given any thought to how we might accomplish this generically?

**Comment by raftmsohani** [comment]
> Looks good to me in general. One thing I didn't see addressed, though, was that Django provides a lot of convenience features through the existing admin form. Django models already have field validators; if we proceed with the implementation as you've outlined, we would need to duplicate these validators in order to have frontend form validation (pre-submission, think field types, ranges, etc). Have you given any thought to how we might accomplish this generically?

@jtimpe  Good catch. I added a Form Metadata and Validation section to clarify this. The intent is not to duplicate Django validators manually in the frontend. Django should remain the source of truth and expose form metadata for migrated admin workflows, derived from the existing model/form layer where possible. The React admin can use that metadata for generic pre-submit validation like required fields, types, etc.. , and of course simple serializable validators.

Anything that cannot be represented losslessly in frontend schema form, such as cross-field validation, custom Python validators, workflow-state rules, or permission-sensitive checks, should remain server-authoritative and be returned as field/non-field errors from Django. So the frontend gets better UX for common mistakes, but Django still owns final validation before persistence/audit behavior. I also added validation drift as an explicit risk/mitigation.

File: docs/Technical-Documentation/Architecture-Decision-Record/023-react-admin-console.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+Django admin is currently used for key administrative workflows in TDP but creates issues at scale:
+
+- Data-heavy pages degrade as records and filters increase.
+- Customization is expensive and tightly coupled to Django admin internals.
+- Admin UX patterns are inconsistent with the React user-facing experience.
+- Section 508 and USWDS consistency require ongoing workaround effort in the Django template layer.
+- Engineering ownership is split across mismatched UI stacks (React for users, Django templates for admins).
+
+These constraints motivate replacing Django admin with a React-based admin console. The decision is how to host and structure that console: inside the existing CRA frontend, or as a standalone Next.js application.
```
Comment (elipe17) [inline docs/Technical-Documentation/Architecture-Decision-Record/023-react-admin-console.md]:
```suggestion
These constraints motivate replacing Django admin with a React-based admin console. The decision is how to host and structure that console: inside the existing Create React App (CRA) frontend, or as a standalone Next.js application.
```

File: docs/Technical-Documentation/React-Admin-Console-Architecture.md [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+└──────────────────────────────────────────────────────────────────┘
+                               │
+                               ▼
+┌──────────────────────────────────────────────────────────────────┐
+│                         PostgreSQL                               │
+└──────────────────────────────────────────────────────────────────┘
+
+
+┌──────────────────────────────────────────────────────────────────┐
+│                          User Browser                            │
```
Comment (elipe17) [inline docs/Technical-Documentation/React-Admin-Console-Architecture.md:127]: Can we move these diagrams to be side by side or in separate triple tick section? These look like they feed each other which is a little confusing

File: docs/Technical-Documentation/React-Admin-Console-Architecture.md [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+- Direct database access from Next.js is out of scope.
+- Any exception to direct DB access requires separate architecture review.
+
+Data access patterns by workflow:
+
+- Read-heavy pages (lists/tables): server-rendered with server-side filtering and pagination.
+- Workflow mutations (approve/reparse/update flags): explicit mutation endpoints with Django-side audit and policy enforcement.
+- Large dataset navigation: paginated or cursor-based API interactions, not client-side bulk loading.
+
+### BFF shaping vs. pass-through pattern
```
Comment (elipe17) [inline docs/Technical-Documentation/React-Admin-Console-Architecture.md]: Can you expand the BFF acronym (backend for frontend) before using it

File: docs/Technical-Documentation/React-Admin-Console-Architecture.md [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+This preserves Django's validation ownership while still giving admins pre-submission feedback for common mistakes. It also gives the migration a practical rule: no Django admin form is retired until its React replacement handles server-returned field and non-field errors and has parity for the generic metadata-supported validations.
+
+---
+
+## Authentication and Authorization
+
+### Authentication model
+
+- Reuse existing Django session-based authentication.
```
Comment (elipe17) [inline docs/Technical-Documentation/React-Admin-Console-Architecture.md]: This feels like a security concern, or like we could do it better. Currently, `/admin/` is reverse-proxied through nginx to the DAC, so the CRA is not acting as a credential handler. The Next.js admin app changes that. The Next.js app becomes a second privileged server-side component that can receive, validate, and replay admin session state. If the CRA and Next.js admin still share one origin, any XSS on the CRA origin could drive authenticated admin requests without needing to read the cookie, and CSRF protections become more complex once mutations can hit Next.js before Django (e.g. BFF). The document should call out this trust-boundary change explicitly and define cache behavior, provenance/audit forwarding, and admin session lifecycle more concretely... Or, we could leverage Keycloak to help us mitigate this.

Keycloak gives us a cleaner way to handle this while keeping authentication seamless. Instead of treating admin as just another path on the main app host, we could put the Next.js admin console on a separate hostname and register it as a separate Keycloak client (like we will for Grafana). In Cloud.gov, that could be a separate `app.cloud.gov` route like `tdp-admin-raft.app.cloud.gov` for dev environments, and a custom domain via the external domain service for prod; e.g. `admin.tanfdata.acf.hhs.gov`. That preserves SSO, because Keycloak can silently authenticate users who already have a valid SSO session, but it gives us a separate browser origin and separate redirect URIs. That means that n the Django side we can have separate sessions which provides better XSS and CSRF boundaries/safety. To support putting the admin on a separate hostname, we could move away from the current broad `SESSION_COOKIE_DOMAIN` settings (`.app.cloud.gov` / `.acf.hhs.gov`) to host-only cookies by leaving `SESSION_COOKIE_DOMAIN` unset/`None`. That would keep the admin session scoped to the admin host and the non-admin session scoped to the CRA host. The same should be true for CSRF cookies. That seems like a lower-complexity way to preserve seamless SSO while reducing cross-subdomain session sharing. What are your thoughts on this?

Comment (raftmsohani) [reply docs/Technical-Documentation/React-Admin-Console-Architecture.md]: Nice catch and review, I agree with this. I updated the architecture doc to make the trust-boundary change explicit and to avoid treating admin as just another path on the CRA host.

The revised target model is to serve tdp-admin from a separate hostname and register it as a separate Keycloak client. That preserves seamless SSO while giving admin its own browser origin, redirect URIs, cookie/CSRF boundary, and session lifecycle. The doc now also recommends host-only session/CSRF cookies where possible instead of broad .app.cloud.gov / .acf.hhs.gov cookie sharing.

I also added guardrails for Next.js as a privileged server-side component: authenticated admin responses should use Cache-Control: no-store, BFF/server actions are privileged paths, Django remains authoritative for authz/audit persistence, and Next.js-to-Django calls must forward provenance context for audit and incident response.

File: docs/Technical-Documentation/React-Admin-Console-Architecture.md (hunk 2 above)
Comment (raftmsohani) [reply docs/Technical-Documentation/React-Admin-Console-Architecture.md:127]: that is a very fair point...

File: docs/Technical-Documentation/React-Admin-Console-Architecture.md [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+
+The rule of thumb: if a single Django endpoint can serve the view, pass-through. If the view needs to join data that Django doesn't serve in a single response, use BFF shaping — but do not add business logic or authorization checks in the BFF layer.
+
+---
+
+## Form Metadata and Validation
+
+Django remains the source of truth for editable admin forms. The React admin should not manually duplicate Django model or form validators in TypeScript. Instead, Django should expose form metadata for each migrated admin workflow through explicit API endpoints.
+
+For model-backed forms, the backend can derive generic metadata from the same Django form/model layer used for server-side validation:
```
Comment (jtimpe) [inline docs/Technical-Documentation/React-Admin-Console-Architecture.md]: do we have a sense for how this could be done, or does it need to be researched further?

File: docs/Technical-Documentation/React-Admin-Console-Architecture.md [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+| Performance regressions on large datasets | High | Server-driven pagination and filtering from day one |
+| Admin trust-boundary expansion | High | Separate admin hostname, separate Keycloak client, host-only admin cookies, Django-authoritative authz |
+| Auth/session edge-case defects | Medium | Explicit session-expiry, logout, cookie-domain, and CSRF test matrix |
+| Accidental caching of privileged admin data | High | `Cache-Control: no-store` for authenticated admin pages and BFF responses |
+| Frontend/backend validation drift | Medium | Generate generic form metadata from Django form/model definitions; keep Django validation authoritative |
+| BFF overgrowth into second backend | Medium | Boundary guardrails in design and review; pass-through as default pattern |
+| Operational overhead from third app | Medium | Reuse existing deployment and monitoring practices |
+| Accessibility drift | Medium | USWDS conformance plus automated checks in CI |
+
+---
```
Comment (jtimpe) [inline docs/Technical-Documentation/React-Admin-Console-Architecture.md:500]: another thing i might add to this is some info about testing - historically we haven't e2e tested the admin page for a few reasons
- it's hard/weird for cypress to access the admin console and page elements
- the existing admin is well tested by virtue of being a django app
- the admin pages are lower priority because they're not user facing

the last point may or may not still be factor, but the first two points should be resolved by this new admin console. it also nicely enables #3465 for existing tests

**Review by jtimpe — COMMENTED** [review]
Some final thoughts - let me know if you want to address them or not (i'm good either way) then i'm happy to approve!

File: docs/Technical-Documentation/React-Admin-Console-Architecture.md (hunk 5 above)
Comment (elipe17) [reply docs/Technical-Documentation/React-Admin-Console-Architecture.md]: I agree with Jan on this. I am wondering how we should expect to architect forms from the frontend and backend perspective. What kind of patterns can we identify to drive generic forms, form builders, etc... There will be a lot of re-used components on the admin frontend and it would be nice to have some guiding design principles to accomplish our goals. Again, doesn't need to call out the UX design, but should call out the technical design.

**Review by elipe17 — CHANGES_REQUESTED** [review]
See my comment on Jan's thread

File: docs/Technical-Documentation/React-Admin-Console-Architecture.md (hunk 5 above)
Comment (raftmsohani) [reply docs/Technical-Documentation/React-Admin-Console-Architecture.md]: Good question. Are you looking for this doc to define the high-level pattern only, or should it include a concrete first-pass technical design for metadata-driven forms?

I assumed the exact metadata contract and frontend form-builder implementation would need a follow-up design spike, but I can expand this section with proposed backend/frontend responsibilities and guardrails if that would be more useful here.

Comment (jtimpe) [reply docs/Technical-Documentation/React-Admin-Console-Architecture.md]: my opinion is the doc should define a high-level pattern - exact details of the contract can be left to the implementer, but i would like to feel confident estimating the ticket without a followup spike. some guiding questions:
* are there any third party libraries that accomplish this for us?
* what backend structure would we have to implement to carry over frontend validation based on the model?
* what does form submission validation look like (more complicated validation than is provided to the frontend)?

i think eric's parser refactor tech memo in #3452 had a good balance of showing how the idea would be implemented without putting in too much detail.

File: docs/Technical-Documentation/React-Admin-Console-Architecture.md (hunk 6 above)
Comment (raftmsohani) [reply docs/Technical-Documentation/React-Admin-Console-Architecture.md:500]: I agree with testing, I mentioned it in technology section, but also question if we have to add more details

File: docs/Technical-Documentation/React-Admin-Console-Architecture.md (hunk 5 above)
Comment (elipe17) [reply docs/Technical-Documentation/React-Admin-Console-Architecture.md]: Agree with Jan here (note havent re-read since latest changes if you've already added this). I think Jan covered everything I would be interested in knowing. It would be awesome if we didn't have to do all the frontend form builder stuff from scratch. But if we did, what does that look like, etc... Small code snippets are helpful especially if they define the interface and not so much the implementation.

**Review by elipe17 — APPROVED** [review]
Once you get another approval, can you add follow on tickets to the existing epic. Only add what you can though since we don't have designs from UX yet.

**Review by jtimpe — APPROVED** [review]
Good work!

**Review by jtimpe — APPROVED** [review]
